### PR TITLE
feat(state_network): implement StateValidator

### PR DESCRIPTION
**Pull request by morph-dev**
### What was wrong?

Validator for the state network wasn't implemented.

### How was it fixed?

Implemented the StateValidator. This includes logic for the recursive gossip.

I added dependency to [my version of the `eth-trie` crate](https://github.com/morph-dev/eth-trie.rs/tree/trin) (I created `trin` branch for this purpose)
- Changes are mostly just making some internal structures and functions public: [see diff](https://github.com/carver/eth-trie.rs/compare/master...morph-dev:eth-trie.rs:trin)
- Still, one function was hard to make public so I made our own copy (with minimal modifications) in: `ethportal-api/src/types/state_trie/utils.rs`

Even with those changes, I'm not very happy with that crate for our purposes. It was designed and optimized for building and updating the Trie, but not for verifying. It was has API from a very high level point of view (doesn't let you interact with internal trie nodes, reason why I needed to make things public), which isn't what we want to do.

I think in the future, we should rewrite that library to be optimized for verification purposes and expose internal structure.

**Note:** PR looks big, but I added a lot of tests. There are a lot of corner cases and I tried to cover most of them (I think there are more lines of codes in tests than in actual code).

### To-Do

- [ ] Clean up commit history and use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

### CODE REVIEW

**Comment by morph-dev** [comment]
TL;DR: I tested it on a local network and it works great!

I made a local network with 3 nodes, and gave 2MB each, which is around 75% of my test data (together they cover entire space).
- I gossiped first 10 blocks (using recursive gossip)
- verified that none of the nodes has all data (manual db queries) 
- queried all trie nodes, and all were returned successfully (and all had correct value)

There was one smaller issue that I observed regarding recursive gossip, but it's in the overlay service and I will address it in a separate PR (which can go either before or after this one).

**Review by njgheorghita — APPROVED** [review]
Bit of a rubber stamp approval, since admittedly I'm not up-to-date on the state network spec. But it looks clean to me and I'm loving all the tests

File: trin-state/src/lib.rs [hunk 1]
```diff
@@ -1,6 +1,3 @@
-#![warn(clippy::unwrap_used)]
-#![warn(clippy::uninlined_format_args)]
```
[the remark this answers is not in the document]
Comment (morph-dev) [reply trin-state/src/lib.rs:2]: I wasn't paying attention. I was thinking that this disables those rules. I added them back. Thanks

File: trin-state/src/validation/validator.rs [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+    error::StateValidationError,
+    recursive_trie_proof::recursive_gossip,
+    trie::{validate_account_state, validate_node_trie_proof},
+};
+
+pub struct StateValidator {
+    pub header_oracle: Arc<RwLock<HeaderOracle>>,
+}
+
+#[async_trait]
```
[the remark this answers is not in the document]
Comment (morph-dev) [reply trin-state/src/validation/validator.rs:30]: The Validator trait (defined in [trin-validation/src/validator.rs](https://github.com/ethereum/trin/blob/666daf60f9d50c715052e504f2779b0b3d229e34/trin-validation/src/validator.rs#L46)) uses `async_trait` and all validators that implement it.
I will look into this in the separate PR.

File: ethportal-api/src/types/state_trie/utils.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
+use eth_trie::node::Node;
+use keccak_hash::{keccak, H256};
+use rlp::RlpStream;
+
+/// Our own implementation for encoding the node.
+///
+/// It is almost identical copy from the `eth_trie` crate, because that one is not reusable.
```
[the remark this answers is not in the document]
Comment (morph-dev) [reply ethportal-api/src/types/state_trie/utils.rs:7]: Not really. There are no tests that test this logic directly (instead they test entire Trie structure and this is tested as part of it).
I would rather not spend time on writing tests for this because:
- this logic is also tested indirectly by state validator logic
- I will want to rewrite library for the trie validation and not use `eth_trie` at all

File: trin-state/src/validation/trie.rs [hunk 4]
```diff
@@ -0,0 +1,719 @@
...
+    Ok(validation_info)
+}
+
+/// Validate the trie proof associated with the account state.
+pub fn validate_account_state(
+    root_hash: Option<H256>,
+    account: &Address,
+    proof: &TrieProof,
+) -> Result<AccountState, StateValidationError> {
+    let mut path = [0u8; 64];
```
[the remark this answers is not in the document]
Comment (morph-dev) [reply trin-state/src/validation/trie.rs:55]: I refactored it. No need anymore.

File: trin-state/src/validation/validator.rs [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+                }
+            }
+            _ => Err(StateValidationError::InvalidContentValueType(
+                "ContractBytecodeKey",
+            )),
+        }
+    }
+
+    async fn get_state_root(&self, _block_hash: H256) -> Option<H256> {
+        // Currently not implemented as history network is not reliable
```
[the remark this answers is not in the document]
Comment (morph-dev) [reply trin-state/src/validation/validator.rs:222]: It would be logged every time state content is offered or retrieved, which would be very spamy.
I added `debug!` instead.

**Comment by morph-dev** [comment]
> Bit of a rubber stamp approval, since admittedly I'm not up-to-date on the state network spec. But it looks clean to me and I'm loving all the tests

Thanks a lot for the review. I know it wasn't easy considering not being up-to-date on the spec.
I will merge it like this, and will fix any followup comments or bugs that we find.
